### ISSUE TITLE
Parallel numerical integration on multiple time axis

### ISSUE THREAD

**Issue by WWXkenmo** [martenlienen/torchode#29]
Thanks for your amazing tool "torchode" which very relevant to my project that I am working on. However, I still has following two questions hope to have your answer.

My archtechture is a bit special. I have N different ODE formula, and each ODE formula would have specific time axis (denote as t_i). Therefore, I need to perform parallel ODE solve over different time axis. And the difficult thing is that you can't sort each axis because it would scramble the data. The initial time t0 and state x0 is known.

So, in my case, the setting is I have (G * N) time matrix, in which G represents the number of variable (time), and N represents the samples. I reshape this matrix into (G * N) * 1 matrix, and use torchode to parallel solve (G * N) batches ODE function, then reshape to get the solved results on different time. The example code as follow:

```
import torch.nn as nn
import torchode as to
import torch

class velocity_encoder(nn.Module):
    
    def __init__(
        self,
        n_int: int = 5,
    ):
        super().__init__()
        self.n_int = n_int
        self.fc1 = nn.Linear(n_int, n_int)
        self.fc2 = nn.Linear(n_int, n_int)
        
    def forward(self,t, u,s):
        
        ## Predict velocity
        du = self.fc1(u)
        ds = self.fc2(s)

        return du, ds

class v_encoder_batch(nn.Module):
    
    def __init__(
        self,
        num_cell: int = 5,
        num_g: int = 5,
    ):
        super().__init__()
        self.num_cell = num_cell
        self.num_g = num_g

    def forward(self,t,y):
        """
        in which x is a reshape matrix: (g*n) * 2
        we first reshape x into two matrix: unspliced (g*n) and spliced (g*n)
        and calculate velocity
        then shape back to the vector: (g*n) * 2
        the batch number in this case is g*n
        """
        u_v = y[:,0]
        s_v = y[:,1]
        u = u_v.reshape(self.num_cell, self.num_g)
        s = s_v.reshape(self.num_cell, self.num_g)
        du, ds = self.v_encoder_class(t, u, s)
    
        ## reshape du and ds
        du = du.reshape(1,self.num_cell*self.num_g)
        ds = ds.reshape(1,self.num_cell*self.num_g)
    
        v = torch.cat([du,ds])
        v = v.T
    
        return v

u = torch.randn(1000,100)
s = torch.randn(1000,100)
x = torch.concatenate([u.reshape(100*1000,1),s.reshape(100*1000,1)],axis = 1)
t = torch.rand(1000,100)*20
t, _ = torch.sort(t, dim=0)

## define velocity
v_encoder = velocity_encoder(n_int = 100)
v_batch = v_encoder_batch(num_cell = 1000, num_g = 100)
v_batch.v_encoder_class = v_encoder

## define initial state
t0 = torch.zeros(t.shape[0]*t.shape[1],1)
x0 = torch.zeros(t.shape[0]*t.shape[1],2)

dt0 = torch.full((t.shape[0]*t.shape[1],), 0.01)
dt0

## reshape t
t_eval = t.reshape(t.shape[0]*t.shape[1],1)
t_eval = torch.cat((t0,t_eval),dim=1)
t_eval

import torchode as to
term = to.ODETerm(v_batch)
step_method = to.Dopri5(term=term)
step_size_controller = to.FixedStepController()
solver = to.AutoDiffAdjoint(step_method, step_size_controller)
sol = solver.solve(to.InitialValueProblem(y0=x0, t_eval=t_eval), dt0 = dt0)
```

Now in the above code, no matter matrix t is sorted or not, it could always gives solution.
Here I have sorted the t, so that I could compare the solution with normal integration routine, which is as follow

```
import torch.nn as nn
import torchode as to
import torch

class velocity_encoder(nn.Module):
    
    def __init__(
        self,
        n_int: int = 5,
    ):
        super().__init__()
        self.n_int = n_int
        self.fc1 = nn.Linear(n_int, n_int)
        self.fc2 = nn.Linear(n_int, n_int)
        
    def forward(self,t, y):
        ## split x into unspliced and spliced readout
        ## x is a matrix with (G*2), in which row is a subgraph (batch)
        #print(y)
        #assert torch.jit.isinstance(args, torch.Tensor)
        #assert torch.jit.isinstance(t, torch.Tensor)
        #assert torch.jit.isinstance(y, torch.Tensor)
        
        ## Predict velocity
        u = y[:,0]
        s = y[:,1]
        
        du = self.fc1(u)
        ds = self.fc2(s)

        du = du.reshape((-1,1))
        ds = ds.reshape((-1,1))

        v = torch.concatenate([du,ds],axis = 1)

        if len(y.shape) == 1:
            v = v.view(-1)

        return v

v_encoder2 = velocity_encoder(n_int = 100)
v_encoder2.fc1 = v_encoder.fc1
v_encoder2.fc2 = v_encoder.fc2
x0 = torch.zeros((100,2))
dt0 = torch.full((100,), 0.01)
t0 = torch.zeros((100,1))

t_eval = torch.cat((t0,t.T),dim=1)
term = to.ODETerm(v_encoder2)
step_method = to.Dopri5(term=term)
step_size_controller = to.FixedStepController()
solver = to.AutoDiffAdjoint(step_method, step_size_controller)
sol2 = solver.solve(to.InitialValueProblem(y0=x0, t_eval=t_eval), dt0 = dt0)
```

However, if look at the solutions of two methods, I observed that two solutions actually have a large difference in some entries, although most entries has small difference.

```
vec = (sol2.ys[:,1:,1].T.reshape(100*1000,1) - sol.ys[:,1:,1]).abs().detach().numpy()
vec = vec.ravel()
vec.mean() #1.0828787
vec.max() #48.40161
```

I am not sure if my solution would be the best way, do you think is make sense?

And another question is that use reshape method would be very very slow. Do your have other way to speed up except JIT?

Best,
Ken

**Comment by martenlienen** [comment]
Hey Ken,
torchode expects `t` to be sorted either increasingly or decreasingly. A random `t` would produce undefined behavior.

`reshape` is a very fast operation because it does not actually move any data round but just creates a new tensor object with new shape information for the same data. If something is very slow, it will surely be a different operation.

Best,
Marten

**Comment by WWXkenmo** [comment]
Yes I know, but in my cases, I can't sort the time matrix t, because it would scramble the data, is that means torchode could not be used in this case?

**Comment by martenlienen** [comment]
Can't you scramble the data in the same way with `torch.argsort`?

**Comment by WWXkenmo** [comment]
My case is a bit complicated, I have a time matrix, and an observation matrix with the same dimension. Each variable would have it specific latent time. Meanwhile, each variable is co-envolve together, their ODE function is coupled. So when I sorted each latent time, that means I have to sort the observation matrix as well, it would scramble the data, and lost their co-envolve information.

**Comment by WWXkenmo** [comment]
Sorry Marten, I know it's a bit hard to explain. Could we possible to schedule some meeting in person and I could show you a bit slides? It's really really important to me. I would be very appreciate for your help! Thanks a lot!

**Comment by martenlienen** [comment]
Why would the following not work?

```python
t = torch.rand((3, 5))
x = torch.rand(3 * 5 * 2).reshape((3, 5, 2))
idx = torch.argsort(t, dim=1)
t_sorted = torch.gather(t, 1, idx)
x_sorted = torch.gather(x, 1, idx.unsqueeze(dim=-1).expand_as(x))
```

**Comment by WWXkenmo** [comment]
Yes, because I don't want x is be sorted, because if x is sorted then it would be wrong to calculate the velocity from this x. Each row of x is not independent with each other, they are coupling when calculate velocity

**Comment by WWXkenmo** [comment]
Thanks Marten for helping me, now is solved
